refactor(core): Use the early event contract instead of the event contract in the bootstrap.

iteriani opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

iteriani committed:
refactor(core): Use early event contract instead of the event contract in bootstrap.

This also fixes an existing bug where we erase the jsaction attribute too early.

Now the event contract binary is 608 bytes :D.

AndrewKushnir commented on packages/core/primitives/event-dispatch/src/register_events.ts:
```diff
@@ -20,6 +22,35 @@ import {EarlyEventContract} from './earlyeventcontract';
  * @returns The `event` contract. This is both assigned to `anyWindow` and returned for testing.
  */
 export function bootstrapEventContract(
```
Do we need to retain this function if we switch over to the `EarlyEventContract`?

iteriani replied:
Hmm I can delete once we remove this file from the copybara

iteriani replied:
actually we cant do that. i need to remove this function once we merge this PR.

iteriani replied:
done

AndrewKushnir reviewed (commented):
@iteriani I think there is some logic missing for the serialization phase: we'd need to exclude all mouse-related events (as that happens today in the `EventContract`) and avoid including them into the contract even if those events were defined in a template. We may also need to do some additional processing (normalizing event names, etc) that happens today in the `EventContract`.

tbondwilkinson commented on packages/core/primitives/event-dispatch/src/earlyeventcontract.ts:
```diff
@@ -34,8 +37,12 @@ export declare interface EarlyJsactionData {
-  constructor() {
-    window._ejsa = {
+  constructor(
+    private readonly container: Window = window,
```
The first parameter is just a general data object - don't type it as Window.

Rename it to something like "replaySink"

iteriani replied:
done

iteriani replied:
done

tbondwilkinson commented on packages/core/primitives/event-dispatch/src/earlyeventcontract.ts:
```diff
@@ -34,8 +37,12 @@ export declare interface EarlyJsactionData {
-    window._ejsa = {
+  constructor(
+    private readonly container: Window = window,
+    private readonly documentElement = window.document.documentElement,
```
This second parameter is the addEventListener root. It's not always documentElement, so don't name it that way and type it as Element

This would more properly be named "container" which is JSAction naming

iteriani replied:
done

tbondwilkinson commented on packages/core/primitives/event-dispatch/src/earlyeventcontract.ts:
```diff
@@ -46,19 +53,19 @@ export class EarlyEventContract {
 
   /**
-   * Installs a list of event types for window.document.documentElement.
+   * Installs a list of event types for documentElement.
```
Update comment

iteriani replied:
done, but this should be reviewed in https://github.com/angular/angular/pull/55586

tbondwilkinson commented on packages/core/src/hydration/event_replay.ts:
```diff
@@ -59,11 +60,17 @@ export function withEventReplay(): Provider[] {
                 const dispatcher = new Dispatcher();
                 setEventReplayer(dispatcher);
-                // Event replay is kicked off as a side-effect of executing this function.
+                eventContract.replayEarlyEvents(container);
```
Move this before new Dispatcher

iteriani replied:
done

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
I would expect tests of the multi-app use case, as it's new behavior.

iteriani replied:
i'd prefer to add tests for this in a followup

tbondwilkinson reviewed (dismissed):
You may also want a test that shows that capture events do not work for EarlyEventContract.

iteriani committed:
refactor(core): Export some more symbols and check for truthiness on event types before adding them.

In some cases, we will be passing in undefined for capture events, so handle this.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

AndrewKushnir commented on packages/core/src/hydration/event_replay.ts:
```diff
@@ -52,7 +55,7 @@ export function withEventReplay(): Provider[] {
         setDisableEventReplayImpl((el: RElement) => {
           if (el.hasAttribute(JSACTION_ATTRIBUTE)) {
-            el.removeAttribute(JSACTION_ATTRIBUTE);
+            removeJsactionQueue.push(el);
```
I'd propose adding a quick comment into the code here to explain why we don't remove the `jsaction` attribute here (that we need it for event replay logic to work and we remove the attribute after we replay all events).

iteriani replied:
oh sure! by the way, i've isolated this change to https://github.com/angular/angular/pull/55696 to make it easier to review

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

AndrewKushnir commented on packages/core/src/hydration/event_replay.ts:
```diff
@@ -73,12 +76,30 @@ export function withEventReplay(): Provider[] {
-                setEventReplayer(dispatcher);
-                // Event replay is kicked off as a side-effect of executing this function.
+              const container = globalThis[CONTRACT_PROPERTY]?.[appId];
+              if (container._ejsa) {
```
I'm wondering if we should avoid using `_ejsa` as a public API. As an alternative, we can:

- either have a few functions to retrieve relevant information (e.g. `getEventTypes`, `getEventTypesWithCapture`, etc)
- or have a function that would return an "expanded" version of the object, for example: `getEarlyJSActionData(container) -> {eventTypes: [...], eventTypesWithCapture: [...],  ...}`

cc @tbondwilkinson for thoughts/feedback as well

AndrewKushnir replied:
We can also have those methods defined in this file and move them over to the JSAction code as a followup PR.

iteriani replied:
done. I think all the other fields (like et, etc) should probably be hidden away as a public API too. I'll do a followup with that one.

AndrewKushnir commented on packages/core/src/hydration/event_replay.ts:
```diff
@@ -73,12 +76,30 @@ export function withEventReplay(): Provider[] {
+                  eventReplayer: (queue) => {
+                    for (const event of queue) {
+                      handleEvent(event);
+                    }
```
Quick question: do we need to clear the queue here?

iteriani replied:
I *think* this queue gets cleared out

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani commented on packages/core/primitives/event-dispatch/src/register_events.ts:
```diff
@@ -54,15 +54,15 @@ export function bootstrapEarlyEventContract(
   const eventContract = new EarlyEventContract(earlyJsactionTracker[field][appId], container);
-  eventContract.addEvents(eventTypes);
-  eventContract.addEvents(captureEventTypes, true);
+  eventTypes && eventContract.addEvents(eventTypes);
+  captureEventTypes && eventContract.addEvents(captureEventTypes, true);
```
This is actually a check if eventTypes and eventCaptureTypes is falsy, so this isn't quite the same.

AndrewKushnir replied:
Ah, sorry, misread this.

alan-agius4 replied:
This also tripped me. Maybe it would be clearer if it's actually an `if` statement.

iteriani replied:
done

iteriani commented on packages/core/primitives/event-dispatch/terser.config.json:
```diff
@@ -7,7 +7,25 @@
+          "h",
+          "c",
+          "etc",
+          "_ejsa"
```
Yeah I actually got these symbols by adding until the test fails, since the test itself evals the binary.

the diff is 60 bytes

alan-agius4 replied:
So the script is now at 4kb with this changes. I also tested using esbuild which resulted in the same size. I'll make a separate PR to switch the build system. 

Well done all in reducing the size so much!

alan-agius4 replied:
Here it is https://github.com/angular/angular/pull/55705

AndrewKushnir replied:
It looks like those symbols are coming from [this code](https://github.com/angular/angular/blob/main/packages/core/primitives/event-dispatch/src/earlyeventcontract.ts#L51) and also from the code in this PR where we access values like `.et`, `.c` from the event contact container (for example [here](https://github.com/angular/angular/pull/55587/files#diff-7c47fb07bd9a28d3333e36e0a631efbe707b9fd4e6ded58de660478ab7dbabdaR90)).

In the first case, I'm wondering if we can avoid creating an object (use an array instead) and for the second case we've discussed that we can come up with a public API that would not rely on those properties. If we can do it, my personal preference is to keep property renaming off for now and re-enable it once we update public APIs (in followup PRs).

WDYT?

iteriani replied:
i'll work on this in a followup. for now i'm reverting this change since it's not buying us nearly as much (700 is pretty good).

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

AndrewKushnir commented:
@iteriani could you please rebase this PR on top of the most recent `main` branch? We've merged PR https://github.com/angular/angular/pull/55708#discussion_r1593015451, which adds size-tracking to the event contract script and it'd likely decrease a lot (which is great!) with the changes from this PR. Once you rebase and push the changes, we'll see what CI says and we can update the size-trackingaccordingly.

iteriani commented:
[TGP](https://fusion2.corp.google.com/presubmit/631168440/OCL:631168440:BASE:631369674:1715115443074:592e47e5/targets)

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani committed:
fixup! refactor(core): Use early event contract instead of the event contract in bootstrap.

iteriani commented on packages/platform-server/src/utils.ts:
```diff
@@ -142,10 +142,25 @@ function insertEventRecordScript(
     const events = Array.from(eventTypesToBeReplayed);
+    const captureEventTypes = [];
+    const eventTypes = [];
+    for (const eventType of events) {
```
done

AndrewKushnir commented on packages/platform-server/src/utils.ts:
```diff
@@ -142,10 +142,33 @@ function insertEventRecordScript(
     const events = Array.from(eventTypesToBeReplayed);
+    const captureEventTypes = [];
+    const eventTypes = [];
+    for (const eventType of events) {
+      if (
+        eventType === 'mouseenter' ||
+        eventType === 'mouseleave' ||
+        eventType === 'pointerenter' ||
+        eventType === 'pointerleave'
+      ) {
+        continue;
+      }
+      if (
+        eventType === 'focus' ||
+        eventType === 'blur' ||
+        eventType === 'error' ||
+        eventType === 'load' ||
+        eventType === 'toggle'
+      ) {
+        captureEventTypes.push(eventType);
```
nit: just as an idea: we can have a couple helper functions from JSAction code, so that we can keep those lists in sync with how JSAction uses them internally.
```suggestion
    for (const eventType of events) {
      if (isMouseEvent(eventType) {
        continue;
      }
      if (useCapturePhase(eventType)) {
        captureEventTypes.push(eventType);
```

AndrewKushnir reviewed (approved):
@iteriani thanks for addressing the feedback!

I've left a comment, but it's **not** blocking, we can do it in a followup PR (it may also require some changes from the JSAction side).

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: fw-core, primitives, fw-platform-server, public-api

iteriani commented:
[TGP](https://fusion2.corp.google.com/presubmit/631168440/OCL:631168440:BASE:631817440:1715183897618:c641aebd;groups=PossiblyNewlyFailing/targets)

thePunderWoman commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -54,9 +54,9 @@
   "platform-server-hydration/browser": {
     "uncompressed": {
-      "main": 200584,
+      "main": 207890,
```
I'm confused why main jumped so much. I would have thought this only would impact the event-dispatch size. Any idea what's up there?

iteriani replied:
we moved logic that was in the event contract to the main bundle.

thePunderWoman reviewed (approved):
reviewed-for: size-tracking

atscott commented:
This PR was merged into the repository by commit 28fb385eec03f64b624a9b67d6d8147b6872a939.
